add label_image for tflite

freedomtan opened the pull request:
label_image for TensorFlow Lite is inspired by TensorFlow's
[label_image](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/label_image), a command line app to load and run classifier
models.

freedomtan committed:
add label_image for tflite

label_image for TensorFlow Lite is inspired by TensorFlow's
label_image, a command line app to load and run classifier
models.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

freedomtan committed:
update path in label_image.md

s/tool:label_image/examples\/label_image:label_image/

andrehentz commented on tensorflow/contrib/lite/examples/label_image/BUILD:
```diff
@@ -0,0 +1,39 @@
+    deps = [
+        "//tensorflow/contrib/lite/kernels:builtin_ops",
+    ],
+    linkopts = select({
```
This could simply be: 
  linkopts = tflite_linkopts()
and you need to:
  load("//third_party/tensorflow/contrib/lite:build_def.bzl", "tflite_copts")

freedomtan replied:
I guest you mean  `load("//tensorflow/contrib/lite:build_def.bzl", "tflite_linkopts")` instead of `load("//third_party/tensorflow/contrib/lite:build_def.bzl", "tflite_copts")`. That also works. However, `/tensorflow/contrib/lite/kernels:builtin_ops` needs `libm`, so I'll use ` linkopts = tflite_linkopts() + ["-lm"]`

andrehentz commented on tensorflow/contrib/lite/examples/label_image/label_image.cc:
```diff
@@ -0,0 +1,467 @@
+// Returns the top N confidence values over threshold in the provided vector,
+// sorted by confidence in descending order.
+template <class T>
+static void get_top_n(std::unique_ptr<tflite::Interpreter>& interpreter,
```
Could you make this independent of the 'interpreter' and move it to a separate file. Also could you unittest it?

freedomtan replied:
done. will push patches later.

andrehentz commented on tensorflow/contrib/lite/examples/label_image/label_image.cc:
```diff
@@ -0,0 +1,467 @@
+}
+
+template <class T>
+void downsize(std::unique_ptr<tflite::Interpreter>& interpreter, int input,
```
Could you move the image manipulation functions into a separate file and unittest them? (It looks like they could be independent of tflite-proper and easy to unittest)

freedomtan replied:
done. will push patches later.

freedomtan committed:
changed according to review

1. move bitmap related functions to bitmap_helpers.cc
2. move get_top_n to get_top_n.cc
3. add some unit tests

freedomtan committed:
make buildifier happy

freedomtan committed:
add -pie to linkopts

Android 5.0 and later supports only position-independent
executibles (PIEs).

andrehentz commented on tensorflow/contrib/lite/examples/label_image/get_top_n.cc:
```diff
@@ -0,0 +1,71 @@
+namespace tflite {
+namespace label_image {
+
+extern bool input_floating;
```
Let's find a way to avoid these global variables.

andrehentz commented on tensorflow/contrib/lite/examples/label_image/get_top_n.cc:
```diff
@@ -0,0 +1,71 @@
+  std::reverse(top_results->begin(), top_results->end());
+}
+
+// explicit instantiation so that we can use them otherwhere
```
As I comment elsewhere, these template instantiations are better placed in a header file somewhere, as free function whose implementations use the template.

freedomtan commented on tensorflow/contrib/lite/examples/label_image/label_image.cc:
```diff
@@ -0,0 +1,312 @@
+int loop_count = 1;
+
+float input_mean = 127.5f;
+float input_std = 127.5f;
```
does packing them into a struct and pass to when necessary sound good?

andrehentz replied:
Definitely. Thanks!

freedomtan commented on tensorflow/contrib/lite/examples/label_image/BUILD:
```diff
@@ -0,0 +1,63 @@
+    name = "label_image",
+    srcs = ["label_image.cc"],
+    linkopts = tflite_linkopts() + [
+        "-pie",  # Android 5.0 and later supports only PIE
```
argh, yes, thanks, I forget other platforms when adding -pie

freedomtan commented on tensorflow/contrib/lite/examples/label_image/label_image.h:
```diff
@@ -0,0 +1,38 @@
+                  int& channels);
+
+template <class T>
+void downsize(T* out, uint8_t* in, int image_height, int image_width,
```
I don't know to do explicit instantiation without the definition of a template function. Can you give me some hint?

freedomtan committed:
change according to review

* eliminate global variables,
* add omitted {}s fro some long single-line statement,
* separate headers,
* replace lena.bmp with grace_hopper.bmp, and
* others

I think I addressed issues except explicit instantiation ones

andrehentz commented on tensorflow/contrib/lite/examples/label_image/label_image.h:
```diff
@@ -0,0 +1,40 @@
+};
+
+// #include "tensorflow/contrib/lite/examples/label_image/bitmap_helpers.h"
+// #include "tensorflow/contrib/lite/examples/label_image/get_top_n.h"
```
would you like to remove those?

freedomtan replied:
argh, sorry about that

andrehentz commented on tensorflow/contrib/lite/examples/label_image/label_image.cc:
```diff
@@ -0,0 +1,300 @@
+}
+
+#if 0
+void RunInference(const string& graph, const string& input_layer_type,
```
remove?

freedomtan replied:
argh, sorry about that

andrehentz commented on tensorflow/contrib/lite/examples/label_image/label_image.h:
```diff
@@ -0,0 +1,40 @@
+using std::string;
+#endif
+
+struct settings {
```
settings -> Settings

andrehentz commented on tensorflow/contrib/lite/examples/label_image/get_top_n.h:
```diff
@@ -0,0 +1,31 @@
+
+template <class T>
+void get_top_n(T* prediction, const int prediction_size,
+               const size_t num_results, const float threshold,
```
you can remove the 'const' from arguments if they are not references or pointers.

andrehentz commented on tensorflow/contrib/lite/examples/label_image/bitmap_helpers.h:
```diff
@@ -0,0 +1,35 @@
+namespace tflite {
+namespace label_image {
+
+#include "tensorflow/contrib/lite/examples/label_image/label_image.h"
```
place the #include line outside the namespaces

freedomtan committed:
clean up accordingly

s/settings/Settings/ and clean up a bit

freedomtan committed:
make buildifier happy

freedomtan committed:
move template functions to *_impl.h

move definitions of template functions into _impl.h so that
we can do explicit instantiation in header files

freedomtan committed:
fix the use of NN API

disable it by default as in the lite/interpreter.cc

drpngx commented:
@freedomtan did you address all review comments?

freedomtan commented:
@drpngx Yes, I think I did. Last batch of changes reflected all issues mentioned by @andrehentz

drpngx commented:
Jenkins, test this please.
